[18.01] Fix scripts/ for YAML/ENV configuration processing

jmchilton opened the pull request:
- Unify configuration parsing within scripts/ when possible.
   - Try to use a common method to setup argparse configuration options.
   - Try to use a common method discover Galaxy properties from ``args``.
   - Try to use a common method to build a Galaxy model and initialize mapping from a config object.
- Most scripts and hopefully all the important ones will now work properly with
    - YAML configuration files (e.g. galaxy.yml) fixing https://github.com/galaxyproject/galaxy/issues/5733.
    - Environment variable configuration semantics (``GALAXY_CONFIG_`` and ``GALAXY_CONFIG_OVERRIDE_``).
    - ``GALAXY_CONFIG_FILE`` to set configuration file location (matching run.sh).
- The above consistency is really important and should improve the usability of these scripts as they are today but also sets us up for implementing ``GALAXY_CONFIG_DIRECTORY`` and ``--galaxy-config-dir`` which may be cool things to have for our imagined ``pip install galaxy`` future.
- Simple integration test cases for over a dozen of the scripts/ including the most important ones.
- Mark several older, less useful scripts as deprecated and note they aren't compatible with "modern configuration". Probably no need to write tests for these and they should likely be dropped.
- Remove 6 even older, even less useful scripts. 2 are broken, 2 haven't been executed in a decade, and another 2 also haven't been used in a decade and don't reflect how purging is done now.
- Swap over OptionParser or ad-hoc sys parsing to ArgumentParser when needed to synchronize argument parsing code for configuration options.
   As a result, a couple new scripts properly accept ``--help`` and print a nice message including ``create_db.py``, and ``secret_decoder_ring.py``.
- Small tweaks to scripts/grt.
   - Drop external dependency on sha256sum so my tests run locally.
   - Drop unused "galaxy_config" configuration option from grt.yml.sample - it doesn't seem to be used.
   - The old ``-c`` option conflicts the new uniform galaxy config file option ``-c`` - so I have changed it to ``-g`` for grt config in both scripts (though only one script uses the Galaxy config). This is technically a breaking change - but I think it is worth it for the consistency. Any objections?

jmchilton committed:
Mark scripts/* that don't work with YAML configuraiton options as deprecated.

jmchilton committed:
Remove unused and unuseful scripts.

jmchilton committed:
Drop data library whoosh/lucene scripts.

We don't set that config option in Galaxy anymore so these are broken anyway.

jmchilton committed:
Remove two scripts - dataset purging hasn't worked this way for a very long time.

jmchilton committed:
Small improvements to scripts/grt/export.py for testing.

- Don't rely on external sha256sum, use Python hashing (add utility to hash_util to enable this).
- Drop unused galaxy_config from grt.yml.sample - it isn't used anywhere as far as I can tell.

jmchilton committed:
Unification of Galaxy configuration parsing for scripts/*

jmchilton commented:
Ping @erasche - made some small changes to ``scripts/grt/*py`` to unify them with the rest of the scripts. I did give you a nice test case in the integration suite for export.py in exchange though. Any chance you can sign off on these changes to the grt stuff?

nsoranzo commented on lib/galaxy/util/script.py:
```diff
@@ -26,30 +26,43 @@ def main(argv=None):
 
 
+def config_file_from_args(args, legacy_config_override=None):
+    # FIXME: you can use galaxy.util.path.extensions for this
```
I think this comment is outdated, `galaxy.util.path.extensions` is used in `find_config_file()`.

jmchilton committed:
Cleanup script integration tests.

nsoranzo commented on lib/galaxy/util/script.py:
```diff
@@ -26,30 +26,43 @@ def main(argv=None):
+    parser.add_argument("-c", "--config-file", "--config",
+                        default=os.environ.get('GALAXY_CONFIG_FILE', None),
+                        help="Galaxy config file (defaults to config/galaxy.ini or config/galaxy.yml)")
+    parser.add_argument("--config-section",
```
I still think this option is not helpful and should not be exposed in all our scripts, seems to be useful only in ` lib/galaxy/model/orm/scripts.py` .

jmchilton replied:
I've made the changes to the methods you requested but I've left this argument in. I think on balance @natefoo and I think it is a good idea and you do not. Are you -1 with it in place and going to force a change or can you live with it here?

I've showed a lot of restraint in not adding ``--config_directory`` and ``--galaxy_root`` as arguments if that helps any 😃 - my impulse is to make this even more adaptable.

nsoranzo replied:
I know that as developers we feel the urge of exposing as many options as possible to the users, but I think that before doing it there should be a use case. Almost no script needed this option till now, so probably there is no use case?

jmchilton replied:
Some of these scripts didn't even allow configuration of the config file, most of them wouldn't respect configuration via environment variables - they would only work under a very narrow range of deployment scenarios. A best practice for deployment (ideal for both package management and containerized deployments) would have strong, clear separation between app directories and configuration directories with different owners and permissions - so one wouldn't place configuration files into config/ directly. That deployment wouldn't work with some of these scripts - it doesn't mean there aren't use cases for doing that - it is the best practice. I'm not arguing that there is a stunning use case here for --config-section just that it is a fallacy that these scripts didn't do something in the past therefore we shouldn't do it or there isn't a use case.

Since galaxy supports the use case of serving multiple instances of galaxy from the same configuration file and some of the scripts feature explicit support for this (galaxy-main, manage_tool_dependencies), all the scripts should. Anything else is a bug since some scripts support a configuration and others don't IMO. How about this - since you aren't advocating gutting the infrastructure that allows this just to reduce the cognitive load on users - what if we set the help to ``argparse.SUPPRESS``. People who want the advanced options have it available to them but we don't populate every help message with it.

nsoranzo replied:
That seems a good compromise, didn't know about `help=argparse.SUPPRESS` ! And sorry for being a pain!

jmchilton committed:
Cleanup incorrect FIXME - spotted by @nsoranzo.

nsoranzo commented on scripts/grt/export.py:
```diff
@@ -167,14 +157,15 @@ def main(argv):
     parser.add_argument('-r', '--report-directory', help='Directory to store reports in',
                         default=os.path.abspath(os.path.join('.', 'reports')))
-    parser.add_argument('-c', '--config', help='Path to GRT config file',
+    parser.add_argument('-g', '--grt-config', help='Path to GRT config file',
```
`args.config` is still used at line 183 below, where it should be updated to `args.grt_config` .

jmchilton replied:
I'll change this - the tests clearly need to be improved also 😠.

jmchilton replied:
Fixing this - caused the test to fail which resulted in more fixes. All in 71b48f1.

nsoranzo commented on lib/galaxy/util/script.py:
```diff
@@ -26,30 +26,43 @@ def main(argv=None):
+def config_file_from_args(args, legacy_config_override=None):
+    app = getattr(args, "app", "galaxy")
+    config_file = legacy_config_override or args.config_file or find_config_file(app)
+    return config_file
```
In order to:
- not require the `--config-section` option
- be usable by Tool Shed scripts, e.g. `scripts/tool_shed/build_ts_whoosh_index.py`

I'd propose to modify the previous 2 functions as below:

```python
def app_properties_from_args(args, legacy_config_override=None, app=None):
    config_file = config_file_from_args(args, legacy_config_override=legacy_config_override, app=app)
    config_section = getattr(args, "config_section", None)
    app_properties = load_app_properties(config_file=config_file, config_section=config_section)
    return app_properties


def config_file_from_args(args, legacy_config_override=None, app=None):
    app = app or getattr(args, "app", "galaxy")
    config_file = legacy_config_override or args.config_file or find_config_file(app)
    return config_file
```

jmchilton committed:
Fix missing ENV information in config parameter args.

jmchilton committed:
Bring in @nsoranzo's requested changes for config_file_from_args, etc.

jmchilton committed:
More grt fixes - spotted by @nsoranzo.

hexylena commented:
> Drop external dependency on sha256sum so my tests run locally.

neat, this looks nicer.

> The old -c option conflicts the new uniform galaxy config file option -c - so I have changed it to -g for grt config in both scripts (though only one script uses the Galaxy config). This is technically a breaking change - but I think it is worth it for the consistency. Any objections?

I think that's fine. Basically no one is using this currently other than freiburg so, feel free to break it.

With @nsoranzo's comments, this looks great. Great to see some testing of this.

jmchilton committed:
Redact help for --config-section option for @nsoranzo.

Kept the string in the file for in-code documentation but hiding it from users.

natefoo commented on lib/galaxy/model/orm/scripts.py:
```diff
@@ -84,23 +105,24 @@ def get_config(argv, cwd=None):
+        if not config_file or get_ext(config_file, ignore='sample') == 'yaml':
+            config_section = database_defaults.get('config_section', None)
+        else:
+            # An .ini file - just let load_app_properties find app:main.
```
This isn't guaranteed to be a PasteDeploy ini file

natefoo commented on lib/galaxy/util/script.py:
```diff
@@ -11,6 +11,19 @@
 DEFAULT_ACTION = None
 
+ARG_HELP_CONFIG_FILE = """
+Galaxy config file (defaults to $GALAXY_ROOT/config/galaxy.yml if that file exists
```
This would be more accurately `./config/galaxy.yml`

jmchilton replied:
I guess it is - I'm not sure how I feel about that. I'll update the docs though.

natefoo commented on scripts/secret_decoder_ring.py:
```diff
@@ -2,50 +2,47 @@
 """
 Script to encode/decode the IDs that galaxy exposes to users and admins.
```
FWIW this functionality is duplicated in helper.py.

jmchilton replied:
I noticed that also - I just kept both for backward compatibility.

hexylena replied:
Huh. I didn't know about that one. I shouldn't've added secret-decoder-ring when it was already existing.

hexylena replied:
Could we remove secret decoder and replace it with a symlink to helper? Any usage of this script is assuredly interactive, so it's not like anyone is going to have flags hard coded into their workflows that they'd have to update.

jmchilton replied:
Can we merge this as is and follow up with that in dev?

hexylena replied:
oh yeah, totally, sorry didn't mean that as a PR blocking comment :)

natefoo reviewed (approved):
Thanks for doing the leg work on these @jmchilton. This really makes things much cleaner.

I can't comment on the comment thread in outdated changes but I just want to mention, can we avoid at all costs flags with underscores in them (e.g. `--config_directory`)?

jmchilton committed:
Touch up script unification PR based on @natefoo comments.

jmchilton commented:
> can we avoid at all costs flags with underscores in them (e.g. --config_directory)?

Yes - this doesn't introduce any new flags with underscores in them.

martenson commented:
Thanks @jmchilton
